README.md
# DEPRECATED!
This module is not supported anymore. The legacy dependencies are too old to maintain.Thank you so far.

You can try [MMM-AssistantMk2](https://github.com/eouia/MMM-AssistantMk2) instead.

## MagicMirror Module: MMM-Assistant

[![MagicMirror2](https://img.shields.io/badge/MagicMirror-2.2.2-lightgray.svg)](https://github.com/MichMich/MagicMirror)
[![DocStatus](https://inch-ci.org/github/eouia/MMM-Assistant.svg?branch=master)](https://inch-ci.org/github/eouia/MMM-Assistant)
[![GitHub last commit](https://img.shields.io/github/last-commit/eouia/MMM-Assistant.svg)](https://github.com/eouia/MMM-Assistant)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/eouia/MMM-Assistant/graphs/commit-activity)
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/eouia/MMM-Assistant.svg)](http://isitmaintained.com/project/eouia/MMM-Assistant "Average time to resolve an issue")
[![Dependency Status](https://beta.gemnasium.com/badges/github.com/eouia/MMM-Assistant.svg)](https://beta.gemnasium.com/projects/github.com/eouia/MMM-Assistant)

A Voice Commander and Google-Assistant for MagicMirror

| STATUS: | Version | Date | Maintained? |
|:------- |:------- |:---- |:----------- |
| Working | `1.0.3` | 2018-03-30 | YES |


#### What is this module doing?

*MMM-Assistant* is a [MagicMirror](https://github.com/MichMich/MagicMirror) (MM) module for using your voice to:  

- control your other *MM* modules
- provide *Google Assistant* functionality
- provide *Alexa* functionality [**WIP**]

---

### Screenshots

There is not much to see since this is a voice based module. But there is a small microphone icon in the modules position.
The icon changes depending on its status, so if you use the wake workds, it will change and also indicate this change by 
playing a "ding". The 2 available wakewords are: **`smart-mirror`** (for MM control) and **`snowboy`** (for GA interaction).
The following icons will appear, depending on the operating *mode* it's in:


| Normal `Listening` | Magic Mirror `Control` | `Google Assistant` | `Amazon Alexa` | `Error` |
|:------------------:|:----------------------:|:------------------:|:--------------:|:-------:| 
| ![Full](./images/Assistant_1.png) | ![Full](./images/Assistant_MM.png) | ![Full](./images/Assistant_GA.png) | ![Full](./images/Assistant_AA.png) | ![Full](./images/Assistant_ProcError.png) |


---

### Installation

There are 5 parts to the installation: 

0. Make sure your RPi OS is not too old...
1. Update *nodejs* and *npm*
2. Install the dependencies
3. Install the Assistant module
4. Setup your *GA* API account
5. Post Installation
6. ~~Configuration~~ 


**(0) To Update your OS**

You should always use the **full** (not `Lite`) version of the Raspbian OS, if you are running on a Raspberry Pi.
The current Raspbian OS version is: `Stretch (9.4)`. You can find your version with:

```bash
uname -a
sudo lsb_release -a
```

If you need to update, use:


```bash
sudo apt-get update 
sudo apt-get upgrade
```


**(1) To Update *nodejs* & *npm***

Before anything, make sure you are using the latest stable release of *nodejs* and *npm*.
They are currently:

```bash
# Check the nodejs and npm versions
node -v && npm -v

v9.10.1
5.8.0
```

If they are very different from this, then update using [these](https://raspberrypi.stackexchange.com/a/77483/17798) instructions.


**(2) To Install the Depdendencies**

```bash
sudo apt-get install libmagic-dev libatlas-base-dev libasound2-dev sox libsox-fmt-all libttspico-utils
```

**(3) To Install the Assistant Module**


```bash
cd ~/MagicMirror/modules
git clone https://github.com/eouia/MMM-Assistant.git
cd MMM-Assistant
date; time npm install         # ~8 min, ~705 packages
```

**(4) Setting up the API Accounts**

*WIP*

:red_circle: Now go to the [Wiki page](https://github.com/eouia/MMM-Assistant/wiki) and follow all the steps in 
`step 3` to create the *Google Assistant* and *Google Cloud Speech* accounts and settings. When done, 
proceed to the post installation below.


**(5) Post Installation**

```bash
cd ~/MagicMirror/modules/MMM-Assistant/
npm install --save-dev electron-rebuild             # ~65 sec
# Do not use the deprecated: --pre-gyp-fix
date; time ./node_modules/.bin/electron-rebuild     # ~25-42 min
```

The last step will take a very long time, as it need to manually compile the [gRPC](https://github.com/grpc/grpc) dependency.  
:warning: This take at least **~25 minutes** on a RPi3, with little or no output. Do not interrupt! :warning:



---

### Module Configuration

To configure the Assistant, you must do the following:

0. Obtain the credentials (as already described above) and add them in the files:
   - `./assets/google-client-secret.json`
   - `./assets/google-access-tokens.json`
   - `./assets/google-private-key.json`   (The OAuth2 certificate file you downloaded and saved.)
1. Add the Module to the global MM `config.js` by copying the content of the file: `./assets/config.txt`.
2. Edit the following fields in the config file:
   - `YOUR_PROJECT_ID`
   - ~~`YOUR_DOWNLOADED_PRIVATE_KEY.json`~~ 


---

Add the module to the MM modules array in the `~/MagicMirror/config/config.js` file by adding the following section.
Here you also need to edit and insert the results you obtained from the Google authorization steps above,
into the fields marked: `YOUR_PROJECT_ID` ~~and `YOUR_DOWNLOADED_PRIVATE_KEY.json` (a filename)~~. 


```javascript
{
    module: 'MMM-Assistant',
    position: 'bottom_left',
    config: {
        assistant: {
            auth: {
                keyFilePath:     "assets/google-client-secret.json", // REQUIRED (Google Assistant API) -- OAuth2 x509 cert
                savedTokensPath: "assets/google-access-tokens.json"  // REQUIRED (Google Assitant API) -- accesss_token & refresh_token
            },
            conversation: {
                lang: 'en-US',
                audio: {
                    encodingIn: "LINEAR16",  // Default. No need to change.
                    sampleRateOut: 16000     // Default. No need to change.
                }
            },
        },
        snowboy: {
            models: [
                {
                    hotwords : "MIRROR"                           // Default model: "MIRROR". (This is not the wake word!)
                    file: "resources/u-models/smart_mirror.umdl", // This file define your MM wake word. (See doc notes.)
                    sensitivity: 0.5,                             // 0.5
                },
                {
                    hotwords : "ASSISTANT"                        // Default model: "ASSISTANT". (This is not the wake word!)
                    file: "resources/u-models/snowboy.umdl",      // This file define your GA wake word. (See doc notes.)
                    sensitivity: 0.5,                             // 0.5
                }
            ]
        },
        record: {
            threshold: 0,                 // Default. No need to change.
            verbose: false,               // Default: true  -- for checking recording status.
            recordProgram: 'rec',         // You can also use 'arecord' or 'sox', but we recommend 'rec'
            silence: 2.0                  // Default. No need to change.
        },
        stt: {
            auth: [{                                             // You can use multiple accounts to save money
                projectId:   'YOUR_PROJECT_ID',                  // REQUIRED (Google Voice API) -- project_id (also shown in key file)
                keyFilename: './assets/google-private-key.json'  // REQUIRED (Google Voice API) -- service_account / private_key
            }],
            request: {
                encoding: 'LINEAR16',     // Default. No need to change.
                sampleRateHertz: 16000,   // Default. No need to change.
                languageCode: 'en-US'     // [en-US]  To set the default GA speech request language.
                                          // See: https://cloud.google.com/speech/docs/languages
            },
        },
        speak: {
            useAlert: true,               // [true]  Enable this to show the understood text of your speech
            language: 'en-US',            // [en-US] To set the default GA speech reply language.
        },
        alias: [{                         // You can use aliases for difficult pronunciation or easy using.
                "help :command" : ["teach me :command", "what is :command"]
        }]
    }
}, // END

```

The main difference with this **full** configuration from the 
[simple](https://github.com/eouia/MMM-Assistant/wiki#simple-version) one, 
is that it uses the `rec` *vox* wrapper for recording, rather than `arecord` since it works better.  
You can change this configuration later when you see that it works.

---

### Available Voice Commands

| Command | Who? | Description |
|:------- |:---- |:----------- |
| **`snowboy`** | GA   | The *Google Assistant* wake word |
| **`smart-mirror`** | MM | The Magic Mirror *command mode* wake word |
| - | - | - |
| help *`command`* | MM | Explain help about a specific *`command`* | 
| list all commands | MM | List all available commands | 
| list all modules | MM | List all available modules | 
| hide all modules | MM | Hide all available modules | 
| show all modules | MM | Show all available modules | 
| say "something" | MM | Use voice to say "`something`"  | 
| reboot | MM | Reboot your RPi | 
| shut down | MM | Shut down your RPi | 



---

### Dependencies (Google API keys)

:red_circle: Unfortunately the **Google Speech API is not free**. But it does allow you `60 minutes` for free per month. 
But every recognition request is considered to last a minimum 15 seconds. So exactly `240 requests` would be available for free. 
Additional requests will then be charged $0.006 per request, if you have a paid accout. Thereafter, the requests will be blocked. 
However, if you create an account for the first time, you will be given about $300 USD in credit, to use within 1 year, for free.

*Therefore it is very important that we get the Amazon Alexa service implementation going, as they are completely free!*


### Dependencies (node/npm)

This module depend on the following *npm* packages:

| npm Package | Version | GitHub |
|:----------- |:------- |:------ |
| [@google-cloud/speech](https://www.npmjs.com/package/@google-cloud/speech) | 1.3.0 | [github](https://github.com/googleapis/nodejs-speech) |
| [google-assistant](https://www.npmjs.com/package/google-assistant) | 0.2.2 | [github](https://github.com/endoplasmic/google-assistant) |
| [grpc](https://www.npmjs.com/package/grpc) | 1.10.0 | [here](https://github.com/grpc/grpc-node/tree/master/packages/grpc-native-core) |
| ~~[node-aplay](https://www.npmjs.com/package/node-aplay)~~ | 1.0.3 | **Deprecated!** | 
| [aplay](https://www.npmjs.com/package/aplay) | 1.2.0 |  [github](https://github.com/roccomuso/node-aplay) | 
| [node-record-lpcm16](https://www.npmjs.com/package/node-record-lpcm16) | 0.3.0 | [github](https://github.com/gillesdemey/node-record-lpcm16) | 
| [snowboy](https://www.npmjs.com/package/snowboy) | 1.2.0 | [github](https://github.com/Kitt-AI/snowboy) | 
| [speaker](https://www.npmjs.com/package/speaker) | 0.4.0 | [github](https://github.com/TooTallNate/node-speaker) |

These are also listed in the `package.json` file and should be installed automatically when using *npm install*.
However, those may require other packages. 

For detailed information, See the [Wiki](https://github.com/eouia/MMM-Assistant/wiki)

---


#### Bugs and Warnings

* Sometimes the GA hangs (issue #25) and the `Error` icon is shown froozen. If this happens,  
  just re-load the MM server page from a remote connection and it will reset.

:information_source: For other bugs, issues, details and updates, please refer to the 
[issue tracker](https://github.com/eouia/MMM-Assistant/issues).


#### Contribution

Feel free to post issues and PR's related to this module. 
For all other or general questions, please refer to the [MagicMirror Forum](https://forum.magicmirror.builders/).

#### Credits

Most grateful thanks to:
* [---](https://github.com/---/) - for clarifying and fixing XXXX

---

#### License

[![GitHub license](https://img.shields.io/github/license/eouia/MMM-Assistant.svg)](https://github.com/eouia/MMM-Assistant/blob/master/LICENSE)

A license to :sparkling_heart:!


CHANGELOG.md
### MMM-Assistant Change Log

All notable changes to this project will be documented in this file.

**Q**: *Why keep a changelog?*  
**A**: [BECAUSE](http://keepachangelog.com/en/1.0.0/)

**Q**: *How?*  
**A**: By carefully describing what was: *added, fixed*, and *changed*.

---

#### [1.1.4] - 2018-04-25

- Added command command to stay awake. This cancels the screen timeout
- Added modulemap can now be a set of modules
- Added command to hide all modules except the selected one. This can be a set from the modulemap so 
    effectively this can be used to show named pages

#### [1.1.3] - 2018-04-20

- Added Google TTS. Many more languages available than pico2tts
  Command conversation can now be in languages other than en-GB,de-DE,es-ES,fr-FR,it-IT like nl-NL
- Added config switch to choose between pico and Google TTS
- Fixed one more bug in GA. Saying "nevermind" just stopped the conversation without returning to listening mode

#### [1.1.2] - 2018-04-20

- Added timeout for screen
- Added screen section in config to specify timeout and screen on/off commands

#### [1.1.1] - 2018-04-19

- Changed handling of config.js so it is now backwards compatible
- fixed some more GA bugs. It should be stable and robust now and handle conversations correctly

#### [1.1.0] - 2018-04-17

- Changed handling of commands in config
- Changed format of models in config.js so an old one will not work (see: assets/config.txt)
- Added commands for screen on and off
- Added commands for showing/hiding a single named module
- Added mapping of spoken module name to actual module name (really helps non-english speakers)
- Added command to play .wav
- Added a single hotword can fire a list of commands and stay in listening mode
  for example, a spoken "wake up" can turn on the screen (like PIR) without activating GA or voice command
- Added model specific confirmation sound added in config (parameter "confirm")
- Added translate/nl.json for dutch language


#### [1.0.3] - 2018-03-31

- Fixed #35 for new GA SDK v0.2.2
- Updated response icons in README
- Updated universal Snowboy models and moved to ./resources/u-models/
- Added response icons for amazon/alexa to ./images directory
- Added Python and Bash scripts to generate Snowboy personal voice models
- Changed google-auth.js scripts location to ./scripts and code
- Changed location of all private files into ./assets + updated relevant code
- Renamed private assets files:  (fixes #42)
    secret.json       --> google-client-secret.json  
    <downloaded>.json --> google-private-key.json  
    tokens.json       --> google-access-tokens.json


#### [1.0.2] - 2018-03-27

Have updated Assistant to use npm:
- google-assistant SDK version 0.2.2.
- @google-cloud/speech  1.3.0
etc.

- Added CHANGELOG
- Added LICENSE
- Added .gitignore
- Added ./docs directory + relevant GitHub content
- Added ./images directory for pictures
- Added ./assets directory for private config,authorization,token files
- Added ./scripts directory for Google & Alexa authrization/token scripts
- Added language support as available in the new GA SDK
- Added French translation
- Changed node_helper.js & MMM-Assistant.js to use GA SDK 22
- Updated README.md with latest instructions
- Updated package.json with updated dependencies and module tags


#### [1.0.1] - 2018-03-XX

- First new commits
- No previous changelog available 


scripts/snowtrain.py
#!/usr/bin/env python
# snowtrain.py -- A Snowboy personal wake-word voice model training script
#----------------------------------------------------------------------------------
# Author:       E:V:A
# FileName:     snowtrain.py
# LastChange:   2018-03-28
# License:      "Give me props!"
# Description:  ...
# Note:         This script is based on the Snowboy repo: training_service.py
# Run With:     ./snowtrain rec1.wav rec2.wav rec3.wav <the-new-word>.pmdl
#----------------------------------------------------------------------------------

import sys
import base64
import requests

def get_wave(fname):
    with open(fname) as infile:
        return base64.b64encode(infile.read())

endpoint = "https://snowboy.kitt.ai/api/v1/train/"

#================  EDIT THE FOLLOWING  ====================
token = ""              # Your Snowbouy API token
hotword_name = "???"    # <the-new-word>
language = "en"         # English!
age_group = "20_29"     # WTF!
gender = "M"            # [M/F] What about trans gender?
microphone = "dick"     # MIC name
#================  END OF EDIT  ===========================

if __name__ == "__main__":
    try:
        [_, wav1, wav2, wav3, out] = sys.argv
    except ValueError:
        print "Usage: %s rec1.wav rec2.wav rec3.wav <the-new-word>.pmdl" % sys.argv[0]
        sys.exit()

    data = {
        "name": hotword_name,
        "language": language,
        "age_group": age_group,
        "gender": gender,
        "microphone": microphone,
        "token": token,
        "voice_samples": [
            {"wave": get_wave(wav1)},
            {"wave": get_wave(wav2)},
            {"wave": get_wave(wav3)}
        ]
    }

    response = requests.post(endpoint, json=data)
    if response.ok:
        with open(out, "w") as outfile:
            outfile.write(response.content)
        print "Saved model to '%s'." % out
    else:
        print "Request failed."
        print response.text



scripts/snowtrain2.sh
#!/usr/bin/env bash
ENDPOINT="https://snowboy.kitt.ai/api/v1/train/"

############# MODIFY THE FOLLOWING #############
TOKEN="??"
#NAME="??"
LANGUAGE="fr"
AGE_GROUP="20_29"
GENDER="M"
MICROPHONE="Jabra speak 410"
############### END OF MODIFY ##################

if [[ "$#" != 1 ]]; then
    printf "Usage: %s out_model_name  (need underscore between words)" $0
    exit
fi

printf "We will record now the 3 audio sample, when you see the record process, say what you want and then press ctrl+C"
printf "Press any key to start the record"
read -n 1 -s

eval "rec -r 16000 -c 1 -b 16 -e signed-integer /tmp/1.wav"
eval "rec -r 16000 -c 1 -b 16 -e signed-integer /tmp/2.wav"
eval "rec -r 16000 -c 1 -b 16 -e signed-integer /tmp/3.wav"

WAV1=`base64 /tmp/1.wav`
WAV2=`base64 /tmp/2.wav`
WAV3=`base64 /tmp/3.wav`
OUTFILE="stt_engines/snowboy/resources/$1.pmdl"
tmp=$1
NAME=$(echo "${tmp//_/ }")


cat <<EOF >data.json
{
    "name": "$NAME",
    "language": "$LANGUAGE",
    "age_group": "$AGE_GROUP",
    "token": "$TOKEN",
    "gender": "$GENDER",
    "microphone": "$MICROPHONE",
    "voice_samples": [
        {"wave": "$WAV1"},
        {"wave": "$WAV2"},
        {"wave": "$WAV3"}
    ]
}
EOF

curl -H "Content-Type: application/json" -X POST -d @data.json $ENDPOINT > $OUTFILE



node_helper.js
/* --------------------------------------------------------------------------
 * Module:       MMM-Assistant
 * FileName:     node_helper.js
 * Author:       eouia
 * License:      MIT
 * Date:         2018-03-31
 * Version:      1.0.3
 * Description:  A MagicMirror module to control your modules
 * Format:       4-space TAB's (no TAB chars), mixed quotes
 *
 * URL:          https://github.com/eouia/MMM-Assistant
 * --------------------------------------------------------------------------
 */

'use strict'

const Sound = require('node-aplay') // Deprecated
//const Sound = require('aplay')
const path = require('path')
const fs = require('fs')
const record = require('node-record-lpcm16')
const textToSpeech = require('@google-cloud/text-to-speech')
const Detector = require('snowboy').Detector
const Models = require('snowboy').Models
const Speaker = require('speaker')
const GoogleAssistant = require('google-assistant')
const speech = require('@google-cloud/speech')
const exec = require('child_process').exec
//const tts = require('picotts')

var NodeHelper = require("node_helper")

module.exports = NodeHelper.create({

  start: function () {
    console.log(this.name + " started");
    this.config = {}
    this.status = 'NOTACTIVATED'
    this.commandAuthIndex = 0
    this.commandAuthMax = 0
    this.pause = new Set()
    this.googleRequestCounter = 1      // Added by E3V3A for Request Count function
  },

  initialize: function (config) {
    this.config = config
    this.config.assistant.auth.keyFilePath     = path.resolve(__dirname, this.config.assistant.auth.keyFilePath)
    this.config.assistant.auth.savedTokensPath = path.resolve(__dirname, this.config.assistant.auth.savedTokensPath)
    this.commandAuthMax = this.config.stt.auth.length
    for(var i=0; i<this.commandAuthMax; i++) {
      this.config.stt.auth[i].keyFilename = path.resolve(__dirname, this.config.stt.auth[i].keyFilename)
    }
    this.sendSocketNotification('MODE', {mode:"INITIALIZED"})
  },

  socketNotificationReceived: function (notification, payload) {
    switch(notification) {
      case 'PAUSE':
        this.pause.add(payload)
        //if (this.pause.size > 0) this.sendSocketNotification('PAUSED')
        break
      case 'RESUME':
        this.pause.delete(payload)
        if (this.pause.size == 0) this.sendSocketNotification('RESUMED')
        break
      case 'CONFIG':
        this.initialize(payload)
        this.status = 'READY'
        break
      case 'HOTWORD_STANDBY':
        if(this.status !== 'HOTWORD_STANDBY') {
          this.status = 'HOTWORD_STANDBY'
          if(this.pause.size == 0) this.activateHotword()
        }
        break
      case 'ACTIVATE_ASSISTANT':
        if (this.status !== 'ACTIVATE_ASSISTANT') {
          this.status = 'ACTIVATE_ASSISTANT'
          if(this.pause.size == 0) this.activateAssistant('ASSISTANT')
        }
        break
      case 'ACTIVATE_COMMAND':
        if (this.status !== 'ACTIVATE_COMMAND') {
          this.status = 'ACTIVATE_COMMAND'
          if(this.config.system.commandRecognition == 'google-cloud-speech') {
            if(this.pause.size == 0) this.activateCommand()
          } else if (this.config.system.commandRecognition == 'google-assistant') {
            if(this.pause.size == 0) this.activateAssistant('COMMAND')
          }
        }
        break
      case 'SPEAK':
        if (this.status !== 'ACTIVATE_SPEAK') {
          this.status = 'ACTIVATE_SPEAK'
          console.log(payload)
          if(this.pause.size == 0) this.activateSpeak(payload.text, payload.option, payload.originalCommand)
        }
        break
      case 'NOTIFY':
        if (payload.notification !== "HOTWORD_STANDBY") {
          if (payload.notification == "PLAY") {
            new Sound(path.resolve(__dirname, payload.parameter)).play();
          } else {
            this.sendSocketNotification("NOTIFY", payload) 
          }
        } else {
          this.status = 'HOTWORD_DETECTED'
          this.sendSocketNotification("HOTWORD_STANDBY")
        }
        break
      case 'EXECUTE':
        execute(payload, function(callback) {
          console.log("[EXECUTE] ", callback)
        })
        break
      case 'REBOOT':
        execute('sudo reboot now', function(callback) {
          console.log(callback)
        })
        break
      case 'SHUTDOWN':
        execute('sudo shutdown -t 1', function(callback) {
          console.log(callback)
        })
        break
      case 'TEST':
        this.test(payload)
        break
      case 'LOG':
        this.consoleLog(payload)
        break
    }
  },

  test: function(test) {
    console.log("TESTING")
  },

  activateSpeak: function(text, commandOption={}, originalCommand = "") {
    var option = {}
    option.language = (typeof commandOption.language !== 'undefined') ? commandOption.language : this.config.speak.language
    option.useAlert = (typeof commandOption.useAlert !== 'undefined') ? commandOption.useAlert : this.config.speak.useAlert
    option.originalCommand = (originalCommand) ? originalCommand : ""
    option.auth = this.config.stt.auth[0]
    option.useGoogle = this.config.speak.useGoogle

    // Use the small footprint Text-to-Speech (TTS): pico2wave
    var commandTmpl = 'pico2wave -l "{{lang}}" -w {{file}} "{{text}}" && aplay {{file}}' 
    var commandGetSpeech = 'pico2wave -l "{{lang}}" -w {{file}} "{{text}}"'
    var commandSpeak = 'aplay {{file}}'

    function getTmpFile() {
        var random = Math.random().toString(36).slice(2),
        path = '/tmp/' + random + '.wav'
        return (!fs.existsSync(path)) ? path : getTmpFile()
    }

    function say(text, lang, cb) {
      text = (text) ? text.trim() : ""
      text = text.replace(/<[^>]*>/g, "")
      text = text.replace(/\"/g, "'")
      text = text.trim()

      var file = getTmpFile()
      var command = ""

      if (option.useGoogle) {
        let client = new textToSpeech.TextToSpeechClient(option.auth)

        const request = {
          input: {text: text},
          voice: {languageCode: lang, ssmlGender: 'NEUTRAL'},
          audioConfig: {audioEncoding: 'LINEAR16'},
        }

        client.synthesizeSpeech(request, (err, response) => {
          if (err) {
            console.error('ERROR:', err);
            return;
          }
          fs.writeFile(file, response.audioContent, 'binary', err => {
            if (err) {
              console.error('ERROR:', err);
              return;
            }
            command = commandSpeak.replace(/\{\{file\}\}/g, file)
            exec(command, function(err) {
              cb && cb(err)
              fs.unlink(file, ()=>{})
            })
          })
        })
      } else {
        command = commandTmpl.replace('{{lang}}', lang).replace('{{text}}', text).replace(/\{\{file\}\}/g, file)
        exec(command, function(err) {
            cb && cb(err)
            fs.unlink(file, ()=>{})
        }) 
      }
    }

    this.sendSocketNotification('MODE', {mode:'SPEAK_STARTED', useAlert:option.useAlert, originalCommand:option.originalCommand, text:text})
    say(text, option.language, (err) => {
      if (!err) {
        console.log("[ASSTNT] Speak: ", text)
        this.sendSocketNotification('MODE', {mode:'SPEAK_ENDED', useAlert:option.useAlert})
        if (this.pause.size > 0) {
          this.sendSocketNotification('PAUSED')
        }
      } else {
        console.log("[ASSTNT] Speak Error", err)
        this.sendSocketNotification('MODE', {mode:'SPEAK_ENDED', useAlert:option.useAlert})
      }
    })
  },

  consoleLog: function(payload) {
  // helper for logging via notification
  // USAGE: sendSocketNotification("LOG", {title: "", message: ""})
    console.log(payload.title, payload.message)
  },

  activateHotword: function() {
    console.log('[ASSTNT] Snowboy Activated')
    this.sendSocketNotification('MODE', {mode:'HOTWORD_STARTED'})
    new Sound(path.resolve(__dirname, 'resources/ding.wav')).play();

    var models = new Models();
    this.config.snowboy.models.forEach((model)=>{
      model.file = path.resolve(__dirname, model.file)
      models.add(model)
    })
    var mic = record.start(this.config.record)
    var detector = new Detector({
      resource: path.resolve(__dirname, "resources/common.res"),
      models: models,
      audioGain: 2.0
    })

    detector.on('silence', ()=>{
      if (this.pause.size > 0) {
        record.stop()
        this.sendSocketNotification('PAUSED')
        return
      }
    })

    detector.on('sound', (buffer)=>{
      if (this.pause.size > 0) {
        record.stop()
        this.sendSocketNotification('PAUSED')
        return
      }
    })

    detector.on('error', (err)=>{
      console.log('[ASSTNT] Detector Error', err)
      record.stop()
      this.sendSocketNotification('ERROR', 'DETECTOR')
      return
    })

    detector.on('hotword', (index, hotword, buffer)=>{
      record.stop()
      var confirm = (typeof this.config.snowboy.models[index-1].confirm !== 'undefined') ?  this.config.snowboy.models[index-1].confirm : 'resources/dong.wav'
      new Sound(path.resolve(__dirname, confirm)).play()
      this.sendSocketNotification('HOTWORD_DETECTED', {hotword:hotword, index:index})
      this.sendSocketNotification('MODE', {mode:'HOTWORD_DETECTED'})
      if (this.pause.size > 0) this.sendSocketNotification('PAUSED')
      return
    })

    mic.pipe(detector);
  },

  activateAssistant: function(mode = 'ASSISTANT') {
    console.log('[ASSTNT] GA Activated')

    var endOfSpeech = false
    var gRQC = this.googleRequestCounter // Added by E3V3A
    var transcription = ""
    this.sendSocketNotification('MODE', {mode:'ASSISTANT_STARTED'})
    const assistant = new GoogleAssistant(this.config.assistant.auth)

    const startConversation = (conversation) => {
      //console.log('Say something!');

      let spokenResponseLength = 0;
      let speakerOpenTime = 0;
      let speakerTimer;
      let openMicAgain = false;

      // This is based on:
      //    ./node_modules/google-assistant/examples/mic-speaker.js
      //    ./node_modules/google-assistant/examples/speaker-helper.js
      conversation
        // send the audio buffer to the speaker
        .on('audio-data', (data) => {
		  const now = new Date().getTime()

                  speaker.write(data)

		  // kill the speaker after enough data has been sent to it and then let it flush out
		  spokenResponseLength += data.length
		  const audioTime = spokenResponseLength / (24000 * 16 / 8) * 1000
		  clearTimeout(speakerTimer)
		  speakerTimer = setTimeout(() => {
                    if (endOfSpeech) {  // if spech.end was already called we notify here
                      this.sendSocketNotification('ASSISTANT_FINISHED', mode)
                    }
                    endOfSpeech = true
                    speaker.end()
		  }, (audioTime - Math.max(0, now - speakerOpenTime)) + 500)
        })
        // done speaking, close the mic
        .on('end-of-utterance', () => { record.stop() })
        // show each word on console as they are understood, while we say it
        .on('transcription', (text) => {
            endOfSpeech = false
//            console.log("[VOX] GA Transcription: ", text)  // show entire JS object
            //---------------------------------------------------------------
            // For account/billing purposes:
            // We check if the transcription is complete and update the request
            // counter by looking for: "done: true". This should probably be
            // moved to MMM-Assistant.
            //---------------------------------------------------------------
            if (text.done)  {
                gRQC += 1
                console.log("[VOX] GA Transcription: ", text.transcription)
                console.log("[VOX] GA RQC: ", gRQC)
            }
            //---------------------------------------------------------------
        })
        // what the assistant answered
        .on('response', text => {
           // another special case. Uttering "nevermind" produces no response
           if (text == "") {
             this.sendSocketNotification('ASSISTANT_FINISHED', mode)
           }
        })
        // if we've requested a volume level change, get the percentage of the new level
        .on('volume-percent', percent => console.log('[VOX] Set Volume [%]: ', percent))
        // the device needs to complete an action
        .on('device-action', action => console.log('[VOX] Action:', action))
        // once the conversation is ended, see if we need to follow up
        .on('ended', (error, continueConversation) => {
          if (error) {
            endOfSpeech = true
            console.log('[ASSTNT] Conversation Ended Error:', error)
            this.sendSocketNotification('ERROR', 'CONVERSATION ENDED')
          } else if (continueConversation) {
            console.log("[ASSTNT - continue conversation]")
            openMicAgain = true
            endOfSpeech = false
          } else if (endOfSpeech) {
            this.sendSocketNotification('ASSISTANT_FINISHED', mode)
          } else {  // weird case where speech.end is called but the  spech is still heard
              endOfSpeech = true
          }
        })
        .on('error', (error) => {
          console.log('[ASSTNT] Conversation Error:', error);
          record.stop()
          speaker.end() // Added by E3V3A: fix attempt for issue #25 --> Need to check for: "Error: Service unavailable"
          this.sendSocketNotification('ERROR', 'CONVERSATION')
//          return // added by E3V3A: Do we also need a return?
        })

      // pass the mic audio to the assistant
      var mic = record.start(this.config.record)
      mic.on('data', data => conversation.write(data));

      // Setup Speaker
      var speaker = new Speaker({
        channels: 1,
        sampleRate: this.config.assistant.conversation.audio.sampleRateOut,
      });
      speaker
        .on('open', () => { 
		  clearTimeout(speakerTimer);
		  spokenResponseLength = 0;
		  speakerOpenTime = new Date().getTime();
		})
        .on('close', () => { if (openMicAgain) assistant.start(this.config.assistant.conversation); });
    };

    // Setup the assistant
    assistant
      .on('ready', () => { assistant.start(this.config.assistant.conversation) })
      .on('started', startConversation)
      .on('error', (error) => {
        console.log('[ASSTNT] Assistant Error:', error)
        record.stop()
        speaker.end()
        this.sendScoketNotification('ERROR', 'ASSISTANT')
      })
  },

  activateCommand: function() {
    this.sendSocketNotification('MODE', {mode:'COMMAND_STARTED'})

    const client = new speech.SpeechClient(this.config.stt.auth[this.commandAuthIndex++])
    if (this.commandAuthIndex >= this.commandAuthMax) this.commandAuthIndex = 0

    const request = {
      config: this.config.stt.request,
      interimResults: false             // If you want interim results, set this to true    // E3V3A: [WTF is this??]
    }
    const recognizeStream = client.streamingRecognize(request)
      .on('error', (err)=>{
        console.log('[ASSTNT] RecognizeStream Error: ', err)
        record.stop()
        this.sendSocketNotification('ERROR', 'RECOGNIZESTREAM')
      })
      .on('data', (data) => {
        this.sendSocketNotification('MODE', {mode:'COMMAND_LISTENED'})
        if ((data.results[0] && data.results[0].alternatives[0])) {
          console.log("[ASSTNT] Command recognized: ", data.results[0].alternatives[0].transcript)
          this.sendSocketNotification('COMMAND',       data.results[0].alternatives[0].transcript)
          record.stop()
        }
        if (this.pause.size > 0) {
          record.stop()
          this.sendSocketNotification('PAUSED')
        }
      })

  // Start Recording and send the microphone input to the Speech API
    record
      .start(this.config.record)
      .on('error', (err)=>{
        console.log("[ASSTNT] Recording Error: ",err)
        record.stop()
        this.sendSocketNotification('ERROR', 'RECORD ERROR')
      })
      .pipe(recognizeStream);
  }
})

//---------------------------------------------------------------------------
//  Helper Functions
//---------------------------------------------------------------------------
function execute(command, callback){
  exec(command, function(error, stdout, stderr){ callback(stdout); });
}

